### PR TITLE
Position accent popup window next to caret

### PR DESCRIPTION
**Pull request by knopp**
https://github.com/flutter/flutter/issues/77545

Depends on https://github.com/flutter/flutter/pull/77608

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by knopp** [comment]
https://github.com/flutter/flutter/issues/30660

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm [hunk 1]
```diff
@@ -187,12 +202,50 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
     // engine since it sent this update, and needs to now be made to match the
     // engine's version of the state.
     [self updateEditState];
+  } else if ([method isEqualToString:@"TextInput.setEditableSizeAndTransform"]) {
+    NSDictionary* state = call.arguments;
+    [self setEditableTransform:state[@"transform"]];
+  } else if ([method isEqualToString:@"TextInput.setCaretRect"]) {
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:208]: Added comments on https://github.com/flutter/flutter/pull/77608. I'd have thought that `TextInput.setMarkedTextRect` should be sufficient here -- despite its name, it should emit the current caret rect when the composing range is empty.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm [hunk 2]
```diff
@@ -377,9 +430,19 @@ - (NSAttributedString*)attributedSubstringForProposedRange:(NSRange)range
 }
 
 - (NSRect)firstRectForCharacterRange:(NSRange)range actualRange:(NSRangePointer)actualRange {
-  // TODO: Implement.
-  // Note: This function can't easily be implemented under the system-message architecture.
-  return CGRectZero;
+  // This only determines position of caret instead of any arbitrary range, but it's enough
+  // to properly position accenet selection popup
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:434]: typo: accent

**Review by cbracken — CHANGES_REQUESTED** [review]
Thanks for sending this! I think if you add a handler for `TextInput.setMarkedTextRect` you should be able to get away without the new message. You'll want to invalidate/re-cache the transformed rect on each update of either the `setMarkedTextRect` or `setEditableSizeAndTransform` message.

**Comment by chinmaygarde** [comment]
@knopp Any updates on @cbracken last comment? If progress is not likely soon, maybe we can close this for now and reopen later?

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm (hunk 1 above)
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:208]: As noted in https://github.com/flutter/flutter/pull/77608, `TextInput.setMarkedTextRect` only works while compositing. So we need another method for caret position.

**Comment by knopp** [comment]
@chinmaygarde, sorry, missed the notifications on this one. I fixed the typo. However this PR does depend on https://github.com/flutter/flutter/pull/77608.

**Comment by chinmaygarde** [comment]
The linked PR has landed. Are there issues with the order in which the commits need to be landed?

**Comment by knopp** [comment]
@cbracken, @chinmaygarde, `TextInput.setMarkedTextRect` doesn't work because it is only valid during compositing. However we need cared position even outside of compositing. This has been discussed in https://github.com/flutter/flutter/pull/77608. I'm not sure what you mean by the linked PR having landed, this depend on https://github.com/flutter/flutter/pull/77608 which hasn't landed yet.

**Comment by cbracken** [comment]
To be clear -- `s/compositing/composing/`. Yep, discussed previously with @knopp and offline with @LongCatIsLooong. 

To summarise discussions elsewhere -- this and setMarkedTextRect cover slightly different but overlapping usecases:

When using a composing IME such as for Chinese, Japanese, Korean input, we need the mark rect since the candidates dropdown menu should remain fixed to the bottom-left of the composing region (depending on platform) until the composing text is committed.

This patch covers non-composing use-cases where an OS-provided accent-selection popup is used to inject an accented character (é, è, ê, etc.) and should move with the caret. I'd actually forgotten this use-case existed as I'd apparently disabled it using `defaults` at some point -- I communicate a fair bit in French but use option-{e,i,u,`} dead-keys for input :(

I did some local testing on macOS, and it appears that when IME composing is active; long-press accent selection is disabled, so there should be no risk of the two behaviours interfering with each other.

**Review by cbracken — APPROVED** [review]
This LGTM. Apologies for the delay -- this fell off my radar while I dealt with a regression to IME composing on macOS after the key event patches landed.

**Comment by GreenLim** [comment]
flutter/flutter#77608 has merged. so these commits can be landed now? it's import for input.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm [hunk 3]
```diff
@@ -403,9 +456,19 @@ - (NSAttributedString*)attributedSubstringForProposedRange:(NSRange)range
...
-  // Note: This function can't easily be implemented under the system-message architecture.
-  return CGRectZero;
+  // This only determines position of caret instead of any arbitrary range, but it's enough
+  // to properly position accent selection popup
+  if (CATransform3DIsAffine(_editableTransform) && !CGRectEqualToRect(_caretRect, CGRectNull)) {
+    CGRect rect =
+        CGRectApplyAffineTransform(_caretRect, CATransform3DGetAffineTransform(_editableTransform));
+
+    // flip and convert to screen coordinates
+    double viewHeight = self.flutterViewController.view.bounds.size.height;
```
Comment (chunhtai) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:466]: self.flutterViewController will be null here ref https://github.com/flutter/engine/pull/23469/files#r612811988

cc @knopp @cbracken

Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm:466]: I noticed that :)

https://github.com/flutter/engine/pull/25524

**Pull request merged**